### ISSUE TITLE
agave-validator revamp

### ISSUE THREAD

**Issue by yihau** [anza-xyz/agave#4082]
#### Motivation

I would like to get `agave-validator` capture environment variables directly so that we can avoid this hardcoded logic:
https://github.com/anza-xyz/agave/blob/8c3a6bb6898e61b0962d6fe9a892dd29b11da097/multinode-demo/bootstrap-validator.sh#L26-L30

currently, we are using clap v2, which has a known bug
ref: https://www.github.com/clap-rs/clap/issues/1476#issuecomment-850292210

it's super painful to upgrade v2 to v4/v5 directly and we also lack sufficient tests to ensure the behavior.

#### Proposed Solution

- Refactor `agave-validator` (the vision for the future: https://github.com/anza-xyz/agave/pull/4083.)
    - reorganize commands and executors
      -  https://github.com/anza-xyz/agave/pull/4084
      - https://github.com/anza-xyz/agave/pull/4761
      - https://github.com/anza-xyz/agave/pull/4762
      - https://github.com/anza-xyz/agave/pull/4771
      - https://github.com/anza-xyz/agave/pull/4774
      - https://github.com/anza-xyz/agave/pull/4775
      - https://github.com/anza-xyz/agave/pull/4776
      - https://github.com/anza-xyz/agave/pull/4781
      - https://github.com/anza-xyz/agave/pull/4782
      - https://github.com/anza-xyz/agave/pull/4783
      - https://github.com/anza-xyz/agave/pull/4784
      - https://github.com/anza-xyz/agave/pull/4799
    - extract clap arg parsing logic and add tests
      - https://github.com/anza-xyz/agave/pull/4893
      - https://github.com/anza-xyz/agave/pull/4902
      - https://github.com/anza-xyz/agave/pull/4903
      - https://github.com/anza-xyz/agave/pull/4922
      - https://github.com/anza-xyz/agave/pull/4923
      - https://github.com/anza-xyz/agave/pull/4929
      - https://github.com/anza-xyz/agave/pull/4930
      - https://github.com/anza-xyz/agave/pull/4990
      - https://github.com/anza-xyz/agave/pull/5246
      - https://github.com/anza-xyz/agave/pull/5253
      - https://github.com/anza-xyz/agave/pull/5254
      - https://github.com/anza-xyz/agave/pull/5255
      - https://github.com/anza-xyz/agave/pull/5256
- Upgrade to clap v3
- Upgrade to clap v4/v5
- Get `agave-validator` capture environment variables directly

**Comment by steviez** [comment]
Before I go review the PR's, I want to ensure I understand the underlying issue / your plan of attack.
- The snippet you linked to shows us having some hard-coded / duplicate logic for parsing our CLI arguments.
- You want to avoid this stuff altogether / get new CLI args supported in this script for free
- The version of CLAP we're on does not support this
- Upgrading to a new version of CLAP has proven to be painful in the past
- You want to break things into separate modules / files and then upgrade CLAP versions

So to confirm, the main goal of this issue & the related PR's would be to get `agave-validator` to a newer version of CLAP ? I guess you remove that snippet from `bootstrap-validator.sh` once we do so, but that seems like a smaller piece than actually getting us to a new CLAP version

**Comment by yihau** [comment]
yeah, the main goal is to upgrade `agave-validator` to a newer version of clap. however the current code structure makes this process hard and painful. imo, there several blockers atm:

1. all args parsing logic is tightly coupled with the core business logic. it's very difficult for us to write some unit tests.
2. there are no existing tests for ensuring `agave-validator` use cases, increasing the risk of breaking existing features during code changes.

I plan to reorganize the code by extracting arg parsing logic to some other places and adding more tests to ensure functionality.

**Comment by yihau** [comment]
btw, here's the vision for the future: https://github.com/anza-xyz/agave/pull/4083.
